### PR TITLE
Add support for side data list

### PR DESCRIPTION
**Pull request by mjh1**
Example ffprobe output with side data:
```
{
    "streams": [
        {
            "index": 0,
            "codec_name": "h264",
            "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
            "profile": "Constrained Baseline",
            "codec_type": "video",
            "codec_tag_string": "avc1",
            "codec_tag": "0x31637661",
            "width": 416,
            "height": 240,
            "coded_width": 416,
            "coded_height": 240,
            "closed_captions": 0,
            "film_grain": 0,
            "has_b_frames": 0,
            "sample_aspect_ratio": "213:208",
            "display_aspect_ratio": "71:40",
            "pix_fmt": "yuv420p",
            "level": 13,
            "chroma_location": "left",
            "field_order": "progressive",
            "refs": 1,
            "is_avc": "true",
            "nal_length_size": "4",
            "id": "0x1",
            "r_frame_rate": "24/1",
            "avg_frame_rate": "602112/24859",
            "time_base": "1/12288",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 24842,
            "duration": "2.021647",
            "bit_rate": "414661",
            "bits_per_raw_sample": "8",
            "nb_frames": "49",
            "extradata_size": 42,
            "disposition": {
                "default": 1
            },
            "tags": {
                "creation_time": "2023-03-30T14:54:40.000000Z",
                "language": "und",
                "handler_name": "Core Media Video",
                "vendor_id": "[0][0][0][0]",
                "encoder": "'avc1'"
            },
            "side_data_list": [
                {
                    "side_data_type": "Display Matrix",
                    "displaymatrix": "\n00000000:       -65536           0           0\n00000001:            0      -65536           0\n00000002:     27918336    15728640  1073741824\n",
                    "rotation": -180
                }
            ]
        },
...
```

I opted for a `map[string]interface{}` as there seem to be 100s of different possible fields and haven't found a definitive list.

With go generics you can write a helper function for retrieving fields with a specific type:
```
func getSideData[T any](s SideData, key string) *T {
	v := s[key]
	switch v.(type) {
	case T:
		res := v.(T)
		return &res
	}
	return nil
}
```

### CODE REVIEW

**Comment by vansante** [comment]
Hmm, this one is tricky, because it is very similar to Tags (just an array of maps), and I don't want to reimplement all the code there.

So my preferred way to implement this would be:

Create a new _private_ struct:

```go
type untypedMap map[string]interface{}
```

then move the functions `GetString`, `GetInt` and `GetFloat` from `Tags` to this new struct.

Then change the definition of Tags to:

```go
type Tags untypedMap
```

and add a new SideData definition:

```go
type SideData untypedMap
```

What do you think?

(Can you also add a unit test perhaps?)

**Comment by mjh1** [comment]
Ah yep that totally makes sense, the only issue with making "aliases" e.g. `type SideData untypedMap` is you then have to keep casting to access the methods like `GetString` so in client code I guess that wouldn't work at all.

I could've done `type SideData Tags` but then you still have the problem of casting, I guess we could include a duplicate set of wrapper methods which do the casting for client code but I figured it might just make sense to use the Tags type again like I have in https://github.com/vansante/go-ffprobe/pull/44/commits/7a003aa52a7e1f2038d1f65ef28c6a42ca48dff5 

What do you think?

**Comment by vansante** [comment]
Right, that is an unexpected gotcha I did not see coming 😅. You can do something [like this for structs, but this](https://sentry.io/answers/alias-type-definitions/) does not work for maps unfortunately.

One final comment though: Does it make sense to make a `SideDataList` struct containing an array of `[]Tags`with a search function so you can quickly query by `side_data_type` (does that field exists for all sidedata and is it unique?) ?

File: ffprobe_test.go [hunk 1]
```diff
@@ -137,3 +137,35 @@ func validateData(t *testing.T, data *ProbeData) {
...
+	if len(videoStream.SideDataList) < 1 {
+		t.Error("Side data list was empty")
+	}
+
+	rotation, err := videoStream.SideDataList[0].GetString("rotation")
+	if err != nil {
+		t.Errorf("Error getting rotation field: %v", err)
+	}
+	if rotation != "-180" {
+		t.Errorf("Expected rotation to be \"-180\", got \"%s\"", rotation)
```
Comment (vansante) [inline ffprobe_test.go]: Maybe use the ` string quotes instead so we can drop the backslashes

File: ffprobe_test.go [hunk 2]
```diff
@@ -137,3 +137,35 @@ func validateData(t *testing.T, data *ProbeData) {
...
+	videoStream := data.FirstVideoStream()
+	if videoStream == nil {
+		t.Error("Video Stream was nil")
+	}
+
+	if len(videoStream.SideDataList) < 1 {
+		t.Error("Side data list was empty")
+	}
+
+	rotation, err := videoStream.SideDataList[0].GetString("rotation")
```
Comment (vansante) [inline ffprobe_test.go]: Rotation is an int field right? So maybe use `GetInt` ?

**Review by vansante — COMMENTED** [review]
Some small nitpicks

File: ffprobe_test.go (hunk 2 above)
Comment (mjh1) [reply ffprobe_test.go]: Oops yep you're right

**Comment by mjh1** [comment]
> Right, that is an unexpected gotcha I did not see coming 😅. You can do something [like this for structs, but this](https://sentry.io/answers/alias-type-definitions/) does not work for maps unfortunately.
> 
> One final comment though: Does it make sense to make a `SideDataList` struct containing an array of `[]Tags`with a search function so you can quickly query by `side_data_type` (does that field exists for all sidedata and is it unique?) ?

Yeah that's a good idea about the search function, hopefully this is what you meant: https://github.com/vansante/go-ffprobe/pull/44/commits/6f192784e3ce1ff8ee561ca79ee09bff9f2a6e5e

File: sidedata.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+const sideDataTypeField = "side_data_type"
+
+// ErrSideDataNotFound is a sentinel error used when a queried side data does not exist
+var ErrSideDataNotFound = errors.New("side data not found")
+
+// SideDataList is a json data structure to represent stream side data
+type SideDataList []Tags
+
+func (s *SideDataList) GetSideData(sideDataType string) (Tags, error) {
```
Comment (vansante) [inline sidedata.go:15]: Can we get a documentation comment on this (public) function?

**Review by vansante — COMMENTED** [review]
One more thing 😉

File: sidedata.go (hunk 3 above)
Comment (mjh1) [reply sidedata.go:15]: Oh yep sorry 👍

File: ffprobe_test.go [hunk 4]
```diff
@@ -137,3 +137,36 @@ func validateData(t *testing.T, data *ProbeData) {
...
+	}
+
+	data, err := ProbeReader(ctx, fileReader)
+	if err != nil {
+		t.Errorf("Error getting data: %v", err)
+	}
+
+	videoStream := data.FirstVideoStream()
+	if videoStream == nil {
+		t.Error("Video Stream was nil")
```
Comment (vansante) [inline ffprobe_test.go:157]: You might want to place a `return` here to satisfy the linter.

**Review by vansante — APPROVED** [review]

**Pull request merged**